proposals/TP1-xds-transport-next.md

```diff
@@ -0,0 +1,1028 @@
+
+```
+message <T>Collection {
+  repeated CollectionEntry resources = 1;
+}
+```
+
+where `CollectionEntry` is defined:
+
+
+```
+message CollectionEntry {
+  // Inlined resource entry.
+  message InlineEntry {
+    // Optional name to describe the inlined resource. Resource names must
+    // [a-zA-Z0-9_-\./]+ (TODO(htuch): turn this into a PGV constraint once
+    // finalized, probably should be a RFC3986 pchar). This name allows
+    // reference via the #entry directive in ResourceLocator.
+    string name = 1 [(validate.rules).string.pattern = "^[0-9a-zA-Z_\\-\\.~:]+$"];
+
+    // The resource's logical version. It is illegal to have the same named xDS
+    // resource name at a given version with different resource payloads.
+    string version = 2;
+
+    // The resource payload, including type URL.
+    google.protobuf.Any resource = 3;
+  }
+
+  oneof resource_specifier {
+    option (validate.required) = true;
+
+    // A resource locator describing how the member resource is to be located.
+    ResourceLocator locator = 1;
+
+    // The resource is inlined in the list collection.
+    InlineEntry inline_entry = 2;
+  }
+}
+```
+
+A simple example of a listener collection fetch sequence:
+
+1. Client requests `xdstp://some-authority/envoy.config.listener.v3.ListenerCollection/foo.`
+2. Server responds with a list of resources: `[locator:
+   xdstp://some-authority/envoy.config.listener.v3.Listener/bar, locator:
+   xdstp://some-authority/envoy.config.listener.v3.Listener/baz].`
+3. Client requests both
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/bar `and
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/baz.`
+4. Server responds with resources contents for
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/bar `and
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/baz`.
+
+This involves two round-trips. Inlining of resources is supported, which can reduce RTT, i.e.
+
+1. Client requests `xdstp://some-authority/envoy.config.listener.v3.ListenerCollection/foo.`
+2. Server responds with a collection of literal resources inlined in `foo`:
+   `[inline_entry: xdstp://some-authority/envoy.config.listener.v3.Listener/bar,
+   inline_entry: xdstp://some-authority/envoy.config.listener.v3.Listener/baz].`
+
+
+#### Glob
+
+Having explicit collection resources allows delta xDS to provide delta updates
+on individual resources but the resource list itself can become a bottleneck for
+frequently updated large collections of resources, e.g. with 1M endpoints updating
+every 10s.  For this situation, we introduce *glob
+collections*, which allow the `<T>Collection` directory to be elided when
+resources in a collection share a directory structure and a `/*` component is appended to
+the resource path. Continuing the previous example, this now looks like:
+
+1. Client requests
+   `xdstp://some-authority/envoy.config.listener.v3.Listener/foo/*.`
+2. Server responds with resources
```

Comment:
Are we going to support inlining resource for glob?

Reply:
No, since there is nothing to inline into. For a list collection, you can inline into the directory message.